Add cycle detection in digraphs using Hadwick & James (2008) algorithm

jgoldfar opened the pull request:
This PR is meant to complement https://github.com/JuliaGraphs/LightGraphs.jl/pull/509, since Johnson's algorithm (should) produce different results on digraphs with loops. In order for this algorithm to work, some functionality is duplicated between the two PRs.

ref https://github.com/JuliaGraphs/LightGraphs.jl/pull/509#issuecomment-294168875

sbromberger commented on src/digraph-cyclicity-hadwick-james.jl:
```diff
@@ -0,0 +1,97 @@
+function simplecycles_hadwick_james end
+@traitfn function simplecycles_hadwick_james(g::::IsDirected)
+    nvg = nv(g)
+    B = [Int[] for i in 1:nvg]
```
```
T = eltype(g)
B = Vector{Vector{T}}()
for i in vertices(g)
  push!(B, Vector{T}())
end
```

sbromberger commented on src/digraph-cyclicity-hadwick-james.jl:
```diff
@@ -0,0 +1,97 @@
+@traitfn function simplecycles_hadwick_james(g::::IsDirected)
+    nvg = nv(g)
+    B = [Int[] for i in 1:nvg]
+    blocked = [false for i in 1:nvg]
```
`blocked = falses(nvg)`

sbromberger commented on src/digraph-cyclicity-hadwick-james.jl:
```diff
@@ -0,0 +1,97 @@
+Find circuits starting from v1 recursively.
+"""
+function circuit_recursive! end
+@traitfn function circuit_recursive!(g::::IsDirected, v1::Int, v2::Int, blocked::Vector{Bool}, B::Vector{Vector{Int}}, stack::Vector{Int}, cycles::Vector{Vector{Int}})
```
if `v1` and `v2` are meant to refer to vertex indices, they should be `T`, not `Int`. Also, with the above change, `blocked` is now a `BitArray`.

Also, you'll want to type-parameterize `B` with `T` (see above).

sbromberger commented on src/digraph-cyclicity-hadwick-james.jl:
```diff
@@ -0,0 +1,97 @@
+        elseif !blocked[w]
+            if circuit_recursive!(g, v1, w, blocked, B, stack, cycles)
+                f = true
+            end
```
can you move the `f = true` to before line 42 since it's being set in both branches?

sbromberger commented on src/digraph-cyclicity-hadwick-james.jl:
```diff
@@ -0,0 +1,97 @@
+    B = [Int[] for i in 1:nvg]
+    blocked = [false for i in 1:nvg]
+    stack = Int[]
+    cycles = Vector{Vector{Int}}()
```
`cycles = Vector{Vector{T}}()`

sbromberger commented on src/digraph-cyclicity-hadwick-james.jl:
```diff
@@ -0,0 +1,97 @@
+"""
+Simultaneously count and remove occurences of a value `val` in the array `list`.
+"""
+function countAndFilter!{T}(val::T, list::AbstractArray{T})
```
In 0.6, this is written
```
function countAndFilter!(val::T, list::AbstractArray{T}) with T
```

sbromberger commented on test/digraph-cyclicity-hadwick-james.jl:
```diff
@@ -0,0 +1,31 @@
+        [2, 3, 4, 5],
+        [2, 3, 5]
+    ]
+    @test issubset(expected_circuits, hadwick_circuits)
```
Check how we use `make_testdigraphs` in `runtests.jl` to ensure this works on other digraphs with different eltypes.

jgoldfar commented:
I've addressed the comments in the latest commit, but now Travis is broken due to some issue in persistence? I have no idea why that would have broken. Code coverage should be back to 100%!

sbromberger commented:
Yeah, that likely isn't your fault. Can I ask you to help out, though? I'm about to push a major change to master. Can we rebase this on top of the new master? It will likely solve this problem.

jgoldfar commented:
After #585 merges? Sure thing.

sbromberger commented:
OK, I merged the branch into master. Let me update and make sure things are good.

sbromberger commented on src/digraph-cyclicity-hadwick-james.jl:
```diff
@@ -0,0 +1,99 @@
+@doc_str """
+Find circuits (including self-loops) using the algorithm of Hadwick & James
```
Oops. The docstrings have the format
```
"""
    simplecycles_hadwick_james(g)

Find circuits (including self-loops) in graph `g` using the algorithm of Hadwick & James.

### References
- Hadwick & James, "Enumerating Circuits and Loops in Graphs with Self-Arcs and Multiple-Arcs", 2008
```

sbromberger commented on src/digraph-cyclicity-hadwick-james.jl:
```diff
@@ -0,0 +1,99 @@
+
+@doc_str """
+Find circuits starting from v1 recursively.
+"""
```
See docstring example above. Also, no need for `@doc_str` unless you're embedding LaTeX.

jgoldfar replied:
I see. I've never had to use the `@doc` string macro; thought it had to be included because of SimpleTraits. Thanks for the clarification!

jgoldfar commented:
Thanks for the review & help!

sbromberger commented on src/digraph-cyclicity-hadwick-james.jl:
```diff
@@ -0,0 +1,101 @@
+        push!(B, Vector{T}())
+    end
+    blocked = falses(nvg)
+    stack = T[]
```
I will fix this later, but this should be Vector{T}().

sbromberger commented on src/digraph-cyclicity-hadwick-james.jl:
```diff
@@ -0,0 +1,101 @@
+Find circuits starting from v1 recursively.
+"""
+function circuit_recursive! end
+@traitfn function circuit_recursive!{T}(g::::IsDirected, v1::T, v2::T, blocked::BitArray, B::Vector{Vector{T}}, stack::Vector{T}, cycles::Vector{Vector{T}})
```
This should be `T <: Integer`.

sbromberger commented on src/digraph-cyclicity-hadwick-james.jl:
```diff
@@ -0,0 +1,101 @@
+        end
+    end
+    filter!(doFilter, list)
+    return nocc
```
Can we rewrite this entire function to be
```
return length(filter!(v -> v == val, list))
```
?

jgoldfar replied:
I think it would have to be 
```
length(list) - length(filter!(v -> v == val, list))
```
but I wasn't sure how that would affect performance. It's only implemented as it is to avoid calling `count` and `filter!`, but none of the alternatives have been benchmarked; may be faster (and definitely easier to read) to simplify it if you prefer.

jgoldfar replied:
Made a Q&D benchmark; looks like your suggestion can be significantly faster
https://gist.github.com/anonymous/054d3c86b77aeba4afe0bc5e333b4f8e
will update the PR to address these comments.

jgoldfar commented on test/digraph-cyclicity-hadwick-james.jl:
```diff
@@ -1,31 +1,58 @@
 
+    # Almost fully connected DiGraph
+    ex4 = DiGraph(9)
+    for (src, dest) in [(1,2),(1,5),(1,7),(1,8),(2,9),(3,4),(3,6),
```
Credit due to https://github.com/nchaudhr for this test case. "Simple" test graphs don't hit the branch in `unblock!`

sbromberger commented:
I've got it - about to create a PR to your branch.

jgoldfar commented:
Sorry if we've duplicated work, but I think I fixed a couple of the things mentioned in your review.

sbromberger commented:
Crap, I may have overwritten your changes. Can you get on gitter or something? (I tried to imessage you).
